core/amber/src/main/scala/edu/uci/ics/texera/web/resource/CollaborationResource.scala

```diff
@@ -18,26 +32,129 @@ object CollaborationResource {
 )
 class CollaborationResource extends LazyLogging {
 
-  @OnOpen
-  def myOnOpen(session: Session): Unit = {
-    websocketSessionMap += (session.getId -> session)
-  }
+  final val objectMapper = Utils.objectMapper
 
   @OnMessage
   def myOnMsg(senderSession: Session, message: String): Unit = {
-    for (sessionId <- websocketSessionMap.keySet) {
-      // only send to other sessions, not the session that sent the message
-      val session = websocketSessionMap(sessionId)
-      if (session != senderSession) {
-        websocketSessionMap(sessionId).getAsyncRemote.sendText(message)
-      }
+    val request = objectMapper.readValue(message, classOf[CollabWebSocketRequest])
+    val uidOpt = senderSession.getUserProperties.asScala
+      .get(classOf[User].getName)
+      .map(_.asInstanceOf[User].getUid)
+    request match {
+      case wIdRequest: InformWIdRequest =>
```

Comment:
I tested another use case, there's a feature "share workflow", user1 can share a workflow to user2 (with write or read access). 
I tested that the collabration only works with the same user, but not different users accessing the same workflow. 
You can try it out by creating two users, then go to the workflow page, click the share button  in a workflow.

Reply:
I think I've heard @Xiao-zhen-Liu mentioned right now it only works with the same user, not the users with sufficient access privilege?

Reply:
Making it work with different users should be a very simple modification. But let's discuss it first.

Reply:
I've changed the code and now it supports multiple users on the same workflow. I also enforced the "readonly" status of workflows shared to other users with read access only, so now a user who is not the owner of a workflow and not having write access will not be able to acquire the lock, like the pic shows:
<img width="391" alt="image" src="https://user-images.githubusercontent.com/36582710/153544900-f3e06e69-4810-48b5-9bd3-981bedd0e777.png">